File: python/common/result.py
# Delta: Closing the Specification Gap — Appendix K
from __future__ import annotations
from dataclasses import dataclass
from typing import Generic, TypeVar, Optional, Callable

T = TypeVar("T")
U = TypeVar("U")


@dataclass(frozen=True)
class Result(Generic[T]):
    """
    Typed result for all business rule outcomes.
    Use instead of raising exceptions for expected failures.

    Usage:
        return Result.success(patient)
        return Result.failure("PATIENT_NOT_FOUND", "Patient not found")
    """

    _value: Optional[T]
    _error_code: Optional[str]
    _error_message: Optional[str]
    _is_success: bool

    @classmethod
    def success(cls, value: T) -> "Result[T]":
        if value is None:
            raise ValueError(
                "success value must not be None - use Result.void_success() for operations "
                "that produce no value"
            )
        return cls(_value=value, _error_code=None,
                   _error_message=None, _is_success=True)

    @classmethod
    def failure(cls, code: str, message: str) -> "Result[T]":
        return cls(_value=None, _error_code=code,
                   _error_message=message, _is_success=False)

    @classmethod
    def void_success(cls) -> "Result[None]":
        return cls(_value=None, _error_code=None,
                   _error_message=None, _is_success=True)

    @property
    def is_success(self) -> bool:
        return self._is_success

    @property
    def is_failure(self) -> bool:
        return not self._is_success

    @property
    def value(self) -> T:
        if not self._is_success:
            raise ValueError(
                f"Result is a failure: [{self._error_code}] {self._error_message}"
            )
        return self._value  # type: ignore[return-value]

    @property
    def error_code(self) -> Optional[str]:
        return self._error_code

    @property
    def error_message(self) -> Optional[str]:
        return self._error_message

    def map(self, func: Callable[[T], U]) -> "Result[U]":
        if self._is_success:
            return Result.success(func(self.value))
        return Result.failure(self._error_code, self._error_message)  # type: ignore

    def __repr__(self) -> str:
        if self._is_success:
            return f"Result.success({self._value!r})"
        return f"Result.failure({self._error_code!r}: {self._error_message!r})"

File: java/common/Result.java
// Delta: Closing the Specification Gap — Chapter 3 / Appendix B
package com.yourcompany.common;

import java.util.Objects;
import java.util.Optional;
import java.util.function.Function;

/**
 * Typed result for all business rule outcomes.
 * Use instead of throwing exceptions for expected failure cases.
 *
 * <pre>
 * // ✓ Correct usage
 * return Result.failure("PATIENT_NOT_FOUND", "Patient not found");
 *
 * // ✗ Incorrect — never throw for business rule failures
 * throw new PatientNotFoundException("Patient not found");
 * </pre>
 */
public final class Result<T> {

    private final boolean success;
    private final T value;
    private final String errorCode;
    private final String errorMessage;

    private Result(boolean success, T value, String errorCode, String errorMessage) {
        this.success      = success;
        this.value        = value;
        this.errorCode    = errorCode;
        this.errorMessage = errorMessage;
    }

    // ── Factories ──────────────────────────────────────────────────────────
    public static <T> Result<T> success(T value) {
        Objects.requireNonNull(value, "success value must not be null");
        return new Result<>(true, value, null, null);
    }

    public static <T> Result<T> failure(String errorCode, String errorMessage) {
        Objects.requireNonNull(errorCode,    "errorCode must not be null");
        Objects.requireNonNull(errorMessage, "errorMessage must not be null");
        return new Result<>(false, null, errorCode, errorMessage);
    }

    /** For void operations — no value on success. */
    public static Result<Void> voidSuccess() {
        return new Result<>(true, null, null, null);
    }

    // ── Accessors ──────────────────────────────────────────────────────────
    public boolean isSuccess() { return success; }
    public boolean isFailure() { return !success; }

    public T getValue() {
        if (!success) throw new IllegalStateException(
            "Result is a failure: [" + errorCode + "] " + errorMessage);
        return value;
    }

    public Optional<T> toOptional() {
        return success ? Optional.ofNullable(value) : Optional.empty();
    }

    public String getErrorCode()    { return errorCode; }
    public String getErrorMessage() { return errorMessage; }

    // ── Transform ──────────────────────────────────────────────────────────
    public <U> Result<U> map(Function<T, U> mapper) {
        return success
            ? Result.success(mapper.apply(value))
            : Result.failure(errorCode, errorMessage);
    }

    @Override
    public String toString() {
        return success
            ? "Result.success(" + value + ")"
            : "Result.failure(" + errorCode + ": " + errorMessage + ")";
    }
}
